### DISCUSSION TITLE
`mypy` throws an error when two pipes are combined

### DISCUSSION THREAD

**Discussion by PetitLepton** [pandas-dev/pandas-stubs#769] [General]
Hi, I am taken aback by the behavior of `mypy`. With the following minimal example
```python
import pandas


def test() -> pandas.DataFrame:
    return pandas.DataFrame().pipe(lambda df: df).pipe(lambda df: df)
```
`mypy` barks at me with
```python
error: <nothing> has no attribute "pipe"  [attr-defined]
```
but a single `pipe` is OK.

Do I have to cast manually the first `pandas.DataFrame().pipe(lambda df: df)`?

I am using `mypy` 1.51, `pandas` 2.0.3 and `pandas-stubs` 2.0.3.230814.

Thanks for your help!

**Comment by twoertwein** [comment]
The issue is that mypy (and pyright)  cannot infer the type of your lambda function but the [definition of pipe](https://github.com/pandas-dev/pandas-stubs/blob/ab5c643b6d6a165ce473e2fc15b1af2aca246d64/pandas-stubs/core/frame.pyi#L1837) depends on the return value of this lambda function.

You could cast, but it would be better to provide type annotations for your lambda function:

```py
from typing import Callable

import pandas as pd

reveal_type(lambda df: df)  # "def (df: Any) -> Any"
reveal_type(pd.DataFrame().pipe(lambda df: df))  # Any

# explicitly declare the type
fun: Callable[[pd.DataFrame], pd.DataFrame] = lambda df: df
reveal_type(pd.DataFrame().pipe(fun).pipe(fun))  # DataFrame

# or use a def
def fun2(df: pd.DataFrame) -> pd.DataFrame:
    return df

reveal_type(pd.DataFrame().pipe(fun2).pipe(fun2))  # DataFrame

```

**Reply by PetitLepton** [reply]
Thanks @twoertwein. I see the issue. I am going to check our code again as all our functions were actually annotated (my minimal example was not in par with that).

**Reply by PetitLepton** [reply]
@twoertwein I just tested the code above. I do not get the same results.
```python
test-mypy.py:5: note: Revealed type is "def (df: Any) -> Any"
test-mypy.py:6: note: Revealed type is "<nothing>"
test-mypy.py:10: error: <nothing> has no attribute "pipe"  [attr-defined]
test-mypy.py:10: note: Revealed type is "Any"
test-mypy.py:16: error: <nothing> has no attribute "pipe"  [attr-defined]
test-mypy.py:16: note: Revealed type is "Any"
```

**Reply by twoertwein** [reply]
It works for me with both the current mypy and pyright (on Fedora with python 3.11 and the current development version of pandas-stbs). I think there wasn't a relevant change since 2.0.3.230814. Does it work for you @Dr-Irv ?

**Reply by PetitLepton** [reply]
Sorry, I think it is on my side. I asked a colleague of mine and he gets the same results as yours. We have the same configuration so I am going to dig into cache issues maybe. Thanks for you help anyway and my apologies for the mess.

**Reply by Dr-Irv** [reply]
Try `mypy --no-incremental` to see if the cache is the issue.

**Reply by PetitLepton** [reply]
Nope. Same result. I am going to create a fresh virtual environment.

**Reply by PetitLepton** [reply]
All right. With a fresh installation, it works as expected. I will dig into that. Thanks a lot!

**Reply by PetitLepton** [reply]
@twoertwein and @Dr-Irv I eventually found the reason. Fresh install works. It turns out that a `mypy` plugin I was using was imposing a lower version of `pandas-stubs`. Without it, it works as expected. Thanks for your help!

**Reply by twoertwein** [reply]
Out of curiosity, which mypy plugin uses pandas-stubs?

**Reply by PetitLepton** [reply]
`pandera`, you can take a look [here](https://github.com/unionai-oss/pandera/blob/main/setup.py#L19).

**Reply by PetitLepton** [reply]
Note that we are using an older version of the `pandera` package which could explain the issues.

**Comment by Dr-Irv** [comment]
@PetitLepton  OK to close the discussion?

**Reply by PetitLepton** [reply]
Sure. I will check on my side and reopen if necessary. :wink: Thanks!
